### PR TITLE
Add warning for putting React prop types inside oneOf.

### PR DESCRIPTION
**Pull request by Agrosis**
First PR!

Took me a few minutes to figure out I needed `oneOfType` instead. Is this worthy? My reasoning for this was that there isn't much to differentiate between `oneOf` and `oneOfType`, if one or the other wasn't known, as opposed to `number` and `string`.

### CODE REVIEW

**Comment by jimfb** [comment]
Seems good to me.  I'll leave it open for a day to collect any feedback before merging.

**Comment by zpao** [comment]
A couple comments:

First off, the big one: *Technically* accepting PropTypes inside oneOf isn't incorrect. I could come up with a contrived use case where I would. Typically though, you probably didn't mean to pass PropTypes in here. So there's the question of whether or not this should be an error. Maybe it should be a warning saying you *probably* didn't mean to use it like this.

Second. As is, you can have validation pass if you did `oneOf(['foo', PropTypes.string])` and then passed `'foo`' as the prop value since you are checking each expected value in order. However if I passed them into `oneOf` in the opposite order, validation would have failed even though I passed a valid value. We shouldn't do any of this checking in `validate` here - we aren't checking the actual values but are validating that you used `oneOf` correctly. Right now you'll get a runtime error checking the props when somebody uses your component. What you really want is an error when you define your component. We already have a check a few lines up doing another usage check, I would do this sort of thing with that code.

Third: Regardless of what we do here, we need tests for a change like this before we can merge it

**Comment by zpao** [comment]
> What you really want is an error when you define your component. We already have a check a few lines up doing another usage check, I would do this sort of thing with that code.

Well, I misread that code a little bit, we still don't fail early in that case but will just always fail validation and short-circuit a more involved check. I would still prefer that happen earlier when defining the component but oh well.

**Comment by Agrosis** [comment]
@zpao, can you link me to one of those usage checks and maybe I can try to make it into that? I'll get some tests in also.

Regarding the error message, it does show up as a 'Warning: ...' in the browser console, but with a red background (in Chrome). You want it to show up with a yellow background, yeah?

**Comment by jimfb** [comment]
@Agrosis I think @zpao's first point needs to be resolved before we tackle the others, since it decides the overall fate of the diff.

Is it reasonable to expect a component to expect proptypes as a proptype?  My intuition is that it's fine to say this is not allowed (people can write their own validator if they really want that), because it's such an edge case compared to the probability that it's a usage mistake.  However, that should be decided before you invest too much work into this diff.

**Comment by zpao** [comment]
A few lines above where your PR starts there a check to see if the argument to `oneOf` was an array. https://github.com/facebook/react/blob/master/src/isomorphic/classic/types/ReactPropTypes.js#L205-L211

You could do something similar.

> Regarding the error message, it does show up as a 'Warning: ...' in the browser console, but with a red background (in Chrome). You want it to show up with a yellow background, yeah?

Well, it's more about when the warning comes. Since PropTypes never actually throw (it's soft validation, we just warn and tell you that you used an invalid value but continue execution), it's about the timing of the warning. In an ideal world, I would want the warning to show up when calling `PropTypes.oneOf(...)` (which happens when defining the component class). However nothing does that so you as a component author won't know you wrote an invalid proptype until you try to use your component (when React does the typechecks at instantiation time).

Anyway, the point there is that we should ideally move this outside of `validate` so that we don't have a mixed validation happening and it will just "fail" consistently.

Both mechanisms would go through `warning` which is a method we have that does `console.error`, so it will be red in the Chrome console regardless.

**Comment by zpao** [comment]
But yea, maybe we could wind it back and ask how how you concluded that `oneOf` would work the way you thought and why you didn't use `oneOfType` to begin with?

(I really hope that doesn't come across as judgemental in any way - you aren't the first one to make this mistake and that means there's a problem on our end. Maybe it can be solved with code or maybe we just write better documentation to avoid special cases in the code.)

**Comment by Agrosis** [comment]
I'm in favor of at least a warning, because if they do end up sticking PropTypes inside `oneOf`, but they don't give it one, they get the `should be one of [null, null]` because functions can't be serialized to json. Maybe make a special case that throws a warning if they said they want a PropType but didn't actually do so?

**Comment by jimfb** [comment]
@zpao I think the only way we could fix that is to rename it `oneOfExactValues` or something.  When I read `oneOf` in the context of type validation, I would assume I could pass in types.  Just my two cents.

**Comment by Agrosis** [comment]
Oh, yeah! The whole point of this PR was to let them know they might've been using the wrong type. Name change could probably help avoid that altogether.

**Comment by jimfb** [comment]
We would need to warn for a release.

**Comment by zpao** [comment]
If we want to go down the renaming path: `PropTypes.enum` - it's an accepted term for this thing and it's what we already call it inside this file.

FWIW, it's been called `oneOf` forever - `oneOfType` was introduced later and there was some indecision with the name (it was brought up that it was confusing, `any` was one of the original ideas dismissed because it was also confusing).

> When I read PropTypes.oneOf in the context of type validation, I would assume I could pass in types.

I can totally see that too. But I think you would have to see it with the incorrect usage to draw that conclusion OR just never have seen either in use and somehow found a `oneOf` property without seeing `oneOfType` as well.

It's still not clear to me where the miscommunication came in? Are there docs somewhere showing incorrect usage? Is it easier to see `oneOf` in our docs and not see `oneOfType`. I'm trying to understand that side of things so we can improve elsewhere as well.

Anyway, I'm not attached to the `oneOf` name, but I think it probably makes sense not to have an intentionally generic "type checker" that then has opinions about what values you're allowed to use.

**Comment by jimfb** [comment]
I can very easily see how reading through an alphabetical list, you would stop as soon as you saw a function called `PropTypes.oneOf`.

Vaguely remember that there was something called "oneOf".  Type "PropTypes.one" in Sublime/Nuclide, oneOf is the first suggestion, you see it, it looks reasonable, accept it, and keep writing, never having seen or consciously thought about the other suggestions in the list.

**Comment by kumfo** [comment]
英文的看起来的确有点吃力，Hah

**Comment by Agrosis** [comment]
If you do choose `oneOf` and put PropTypes in there, expecting it to work like `oneOfType`, you get back a `should be one of [null, null].`, which isn't very helpful. I think I saw an issue about this a few days ago.

**Comment by gaearon** [comment]
Hey @Agrosis, thank you for the PR! Sorry we didn’t communicate any resolution on this for a long time.

I will close this due to the reasons described in https://github.com/facebook/react/pull/5418#issuecomment-155949060.

I also would like to add that this solution is also not very consistent for another reason. It will warn for `React.PropTypes.oneOf([React.PropTypes.number, React.PropTypes.string])` but not for something like
```js
React.PropTypes.oneOf([
  React.PropTypes.number.isRequired, // .isRequired will give a different proptype
  React.PropTypes.shape({
    x: React.PropTypes.number
  }), // creates a new proptype
  MyCustomPropTypes // is not listed on React.PropTypes
])
```

In other words it “sees” only top-level `React.PropTypes` whereas the proptypes system is meant to work with any custom function and produces new functions in some cases.

I will close for these reasons but please don’t feel discouraged from sending more PRs! In general it’s best to discuss the issue first without diving into the implementation. For example, we can discuss whether we want to add a warning or to rename PropTypes to solve the confusion.

In any case, please know that your contribution is appreciated. Watch out for [good first bugs](https://github.com/facebook/react/issues?q=is%3Aopen+is%3Aissue+label%3A%22good+first+bug%22) as well!

Kudos.
